### PR TITLE
BVS-6943: validate IPv4 address is valid host address

### PR DESCRIPTION
**Pull request by marymanohar**
Reviewer: @andi-bigswitch

### CODE REVIEW

**Comment by bsn-abat** [comment]
Find artifact changes from this pull request at https://github.com/floodlight/loxigen-artifacts/tree/pr-531-jenkins-loxigen_pull_req_artifacts-78

**Comment by bsn-abat** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://jenkins.bigswitch.com/job/loxigen_pull_req_artifacts/78/

**Comment by bsn-abat** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://jenkins.bigswitch.com/job/loxigen_pull_req_docker/85/

File: java_gen/pre-written/src/main/java/org/projectfloodlight/openflow/types/IPv4Address.java [hunk 1]
```diff
@@ -117,6 +117,61 @@ public boolean isMulticast() {
...
+        LINK_LOCAL_NETWORK,
+        TEST_NET,
+        RELAY_ANYCAST_NETWORK,
+        INTERCONNECT_NETWORK,
+        MULTICAST_NETWORK,
+        RESERVED_NETWORK,
+        HOST_ADDRESS
+    }
+
+    public IPAddressType isHostAddress() {
```
Comment (andi-bigswitch) [inline java_gen/pre-written/src/main/java/org/projectfloodlight/openflow/types/IPv4Address.java]: `isHostAddress` implies a boolean return value; this should probably be called `getAddressType`?

File: java_gen/pre-written/src/main/java/org/projectfloodlight/openflow/types/IPv4Address.java [hunk 2]
```diff
@@ -117,6 +117,61 @@ public boolean isMulticast() {
...
+        PUBLIC_DATA_NETWORK,
+        CABLE_TV_NETWORK,
+        INTERNET_HOST_LOOPBACK_NETWORK,
+        LINK_LOCAL_NETWORK,
+        TEST_NET,
+        RELAY_ANYCAST_NETWORK,
+        INTERCONNECT_NETWORK,
+        MULTICAST_NETWORK,
+        RESERVED_NETWORK,
+        HOST_ADDRESS
```
Comment (andi-bigswitch) [inline java_gen/pre-written/src/main/java/org/projectfloodlight/openflow/types/IPv4Address.java]: Are we sure that `HOST_ADDRESS` is a valid terminology here? IMHO, a valid `HostAddress` depends on the network in which it exists; in particular it must be within the network and not the network or broadcast address of that network.

File: java_gen/pre-written/src/main/java/org/projectfloodlight/openflow/types/IPv4Address.java [hunk 3]
```diff
@@ -117,6 +117,61 @@ public boolean isMulticast() {
...
+    }
+
+    public IPAddressType isHostAddress() {
+        if (IPv4AddressWithMask.THIS_NETWORK.contains(this))
+            return IPAddressType.THIS_NETWORK;
+
+        if ((IPv4AddressWithMask.PRIVATE1_NETWORK.contains(this)) ||
```
Comment (andi-bigswitch) [inline java_gen/pre-written/src/main/java/org/projectfloodlight/openflow/types/IPv4Address.java]: This if-then-else cascade might be expensive if executed on lots of addresses would be better to use a trie.

File: java_gen/pre-written/src/main/java/org/projectfloodlight/openflow/types/IPv4Address.java [hunk 4]
```diff
@@ -117,6 +117,61 @@ public boolean isMulticast() {
...
+            (IPv4AddressWithMask.PRIVATE2_NETWORK.contains(this)) ||
+            (IPv4AddressWithMask.PRIVATE3_NETWORK.contains(this)))
+            return IPAddressType.PRIVATE_NETWORK;
```
Comment (andi-bigswitch) [inline java_gen/pre-written/src/main/java/org/projectfloodlight/openflow/types/IPv4Address.java]: With respect to the intended use in BVS-6943 -- note that addresses within the RFC-1918 private networks are of course perfectly adequate to use (as we do in our networks). The same goes for other 'special' networks listed in RFC3330. Is the intended use that a client would check against all the `IPAddressType`s that are acceptable?

File: java_gen/pre-written/src/main/java/org/projectfloodlight/openflow/types/IPv4AddressWithMask.java [hunk 5]
```diff
@@ -8,6 +8,77 @@
...
+     * 192.168.0.0/16       Private-Use Networks                   [RFC1918]
+     * 198.18.0.0/15        Network Interconnect
+     *                      Device Benchmark Testing            [RFC2544]
+     * 223.255.255.0/24     Reserved but subject
+     *                      to allocation                             --
+     * 224.0.0.0/4          Multicast                              [RFC3171]
+     * 240.0.0.0/4          Reserved for Future Use        [RFC1700, page 4]
+
+     */
+    public static final IPAddressWithMask<?> THIS_NETWORK =
```
Comment (andi-bigswitch) [inline java_gen/pre-written/src/main/java/org/projectfloodlight/openflow/types/IPv4AddressWithMask.java]: It's ugly that the related definitions of the networks and the types are split across two classes here. I am wondering if if would be best to put them in a separate class?

**Comment by marymanohar** [comment]
Looks like we don't need to check for all these subnets for the JIRA in question. We just need to check for link-local and loopback addresses. That can be done in BCF itself, in an existing validator function.

**Pull request closed without merging**